README.md
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [@nomadreservations/ngx-stripe](#nomadreservationsngx-stripe)
  - [Features](#features)
  - [Installation](#installation)
  - [Using the library](#using-the-library)
  - [StripeCardComponent](#stripecardcomponent)
    - [RequestPaymentButton](#requestpaymentbutton)
    - [Dynamic Stripe Keys](#dynamic-stripe-keys)
    - [PaymentIntents](#paymentintents)
    - [SCA/3D Secure payments](#sca3d-secure-payments)
  - [Testing](#testing)
  - [Building](#building)
  - [Publishing](#publishing)
  - [Documentation](#documentation)
  - [License](#license)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# @nomadreservations/ngx-stripe

[![version](https://img.shields.io/npm/v/@nomadreservations/ngx-stripe.svg)](https://www.npmjs.com/package/@nomadreservations/ngx-stripe)
[![license](https://img.shields.io/npm/l/express.svg)](https://www.npmjs.com/package/@nomadreservations/ngx-stripe)

@angular wrapper for StripeJS based on module by Ricardo Sánchez Gregorio at https://github.com/richnologies/ngx-stripe

_Note: This version is designed to work with angular v8+ only_

## Features

* Stripe Service
* Lazy script loading
* AoT compliant
* SSR fallback

## Installation

To install this library, run:

```bash
$ npm install @nomadreservations/ngx-stripe --save
or
$ yarn add @nomadreservations/ngx-stripe
```

## Using the library

Import the `NgxStripeModule` into the application

The module takes the same parameters as the global Stripe object. The APIKey and the optional options to use Stripe connect

* apiKey: string
* options: {
  stripeAccount?: string;
}

```typescript
import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';

import { AppComponent } from './app.component';

// Import your library
import { NgxStripeModule } from '@nomadreservations/ngx-stripe';

@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    BrowserModule,
    NgxStripeModule.forRoot('***your-stripe-publishable key***'),
    LibraryModule
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
```

_Note: If you leave the publishable key blank you must set on using StripeSerive.changeKey prior to creating token's or sources_

## StripeCardComponent

As an alternative to the previous example, you could use the StripeCardComponent.

It will make a little bit easier to mount the card.

To fetch the Stripe Element, you could you use either the (change) output, or,
by using a ViewChild, the public method getCard()

//stripe.compnent.html
```html
<ngx-stripe-card [options]="cardOptions" [elementsOptions]="elementsOptions" (change)="cardUpdated($event)" (error)="error = $event"></ngx-stripe-card>
<div class="error">
  {{error?.message}}
</div>
<button (click)="getCardToken()" [disabled]="!complete">Get Card Token</button>
```

//stripe.component.ts
```typescript
import { Component, OnInit, ViewChild, ElementRef } from '@angular/core';
import { FormGroup, FormBuilder, Validators } from "@angular/forms";

import { StripeService, StripeCardComponent, ElementOptions, ElementsOptions } from "@nomadreservations/ngx-stripe";


@Component({
  selector: 'app-stripe-test',
  templateUrl: 'stripe.component.html'
})
export class StripeTestComponent implements OnInit {
  stripeKey = '';
  error: any;
  complete = false;
  element: StripeElement;
  cardOptions: ElementOptions = {
    style: {
      base: {
        iconColor: '#276fd3',
        color: '#31325F',
        lineHeight: '40px',
        fontWeight: 300,
        fontFamily: '"Helvetica Neue", Helvetica, sans-serif',
        fontSize: '18px',
        '::placeholder': {
          color: '#CFD7E0'
        }
      }
    }
  };

  elementsOptions: ElementsOptions = {
    locale: 'en'
  };

  constructor(
    private _stripe: StripeService
  ) {}

  cardUpdated(result) {
    this.element = result.element;
    this.complete = result.card.complete;
    this.error = undefined;
  }

  keyUpdated() {
    this._stripe.changeKey(this.stripeKey);
  }

  getCardToken() {
    this._stripe.createToken(this.element, {
      name: 'tested_ca',
      address_line1: '123 A Place',
      address_line2: 'Suite 100',
      address_city: 'Irving',
      address_state: 'BC',
      address_zip: 'VOE 1H0',
      address_country: 'CA'
    }).subscribe(result => {
      // Pass token to service for purchase.
      console.log(result);
    });
  }
}
```

### RequestPaymentButton
A payment request button (e.g. apple pay, google chrome payment, etc.) can be enabled with the `ngx-payment-request` component. All of the options [Request Payment Api](https://stripe.com/docs/payment-request-api) options are exposed via the three options sections in the component. 

// request-button.component.html
```html
<div class="container">
  <mat-card>
    Pay: <mat-slider min="1" max="100" step="10" thumbLabel [value]="pay" (valueChange)="updatePay($event)"></mat-slider>
    <ngx-payment-request 
      [options]="requestOptions" 
      [elementsOptions]="elementsOptions" 
      [styles]="styles" 
      (shippingAddressChange)="updateShippingAddress($event)"
      (shippingOptionChange)="updateShippingOption($event)"
      (change)="requestUpdated($event)">
    </ngx-payment-request>

  </mat-card>
```

// request-button.compnent.ts
```typescript
@Component({
  selector: 'app-test',
  templateUrl: 'request-button.component.html'
})
export class RequestButtonComponent implements OnInit {
  @Input set stripeKey(key) {
    this._stripe.changeKey(key);
  }

  public requestOptions: RequestElementOptions = {
    country: 'US',
    currency: 'usd',
    requestPayerName: true,
    requestPayerEmail: true,
    requestPayerPhone: true,
    requestShipping: true,
    total: {
      amount: 10000,
      label: 'Donate to the things'
    }
  };

  public styles: PaymentRequestButtonStyle = {
    type: 'donate',
    theme: 'light',
    height: '64px'
  };

  public elementsOptions: ElementsOptions = {
    locale: 'en'
  };

  public pay = 100;

  constructor(private _stripe: StripeService) { }
  
  public updatePay(amount: number) {
    this.pay = amount;
    this.requestOptions = {
      ...this.requestOptions,
      total: {
        ...this.requestOptions.total,
        amount: amount * 100
      }
    };
  }

  public async requestUpdated(result) {
    const response = await fetch('/charges', {
      method: 'POST',
      body: JSON.stringify({token: result.token.id}),
      headers: {'content-type': 'application/json'},
    });

    if (response.ok) {
      result.complete('success');
    } else {
      result.complete('fail');
    }    
  }

  public updateShippingAddress(result) {
    result.updateWith({
      status: 'success',
      shippingOptions: [
        {
          id: 'free-shipping',
          label: 'Free shipping',
          detail: 'Arrives in 5 to 7 days',
          amount: 0
        },
        {
          id: 'express',
          label: 'Express shipping',
          detail: 'Arrives in 1 to 2 days',
          amount: 1000
        }
      ]
    });
  }

  public updateShippingOption(result) {
    result.updateWith({
      status: 'success'
    });
  }
}

```

### Dynamic Stripe Keys
Given the way an AOT angular build processes provided values if you need to provide a stripe key at runtime you'll want to forgo the forRoot initialization and instead add a factory and provider

```ts

  export function NgxStripeFactory(): string {
    return dynamicLocation.stripePkKey;
  }
  
  @NgModule({
    ...
    imports: []
      ...
      NgxStripeModule.forRoot('whatever key you think should be default'),
      ...
    ]
    providers: [
      {
        provide: STRIPE_PUBLISHABLE_KEY,
        useFactory: NgxStripeFactory
      }
    ]
  })
```

### PaymentIntents
You can use the payment intents api with this service just like with stripe.js the following are exposed via the `StripeService`

```ts
  handleCardSetup(clientSecret: string, el: Element, cardSetupOptions?: SetupIntentData): Promise<SetupIntentResult>;
  handleCardAction(clientSecret: string);
  handleCardPayment(clientSecret: string, el: Element, data: CardPaymentData);
  confirmPaymentIntent(clientSecret: string, el: Element, data: ConfirmIntentData);
  retrievePaymentIntent(clientSecret: string);
  confirmSetupIntent(clientSecret: string, el: Element, data: ConfirmSetupIntentData);
  retrieveSetupIntent(clientSecret: string);
```

### SCA/3D Secure payments

3D Secure payments are enabled by following a [3D Secure Flow](https://stripe.com/docs/payments/3d-secure#example-of-a-3d-secure-2-flow) using the payment intents api's. You will need to setup a proper next action in the payment intent based off of the 3D Secure response from the customers bank/institution. 

Example:
//stripe.compnent.html
```html
<ngx-stripe-card [options]="cardOptions" [elementsOptions]="elementsOptions" (change)="cardUpdated($event)" (error)="error = $event"></ngx-stripe-card>
<div class="error">
  {{error?.message}}
</div>
<button (click)="payNow()" [disabled]="!complete">Pay</button>
```

//stripe.component.ts
```typescript
import { Component, OnInit, ViewChild, ElementRef } from '@angular/core';
import { FormGroup, FormBuilder, Validators } from "@angular/forms";

import { StripeService, StripeCardComponent, ElementOptions, ElementsOptions } from "@nomadreservations/ngx-stripe";


@Component({
  selector: 'app-stripe-test',
  templateUrl: 'stripe.component.html'
})
export class StripeTestComponent implements OnInit {
  stripeKey = '';
  error: any;
  complete = false;
  element: StripeElement;
  cardOptions: ElementOptions = {
    style: {
      base: {
        iconColor: '#276fd3',
        color: '#31325F',
        lineHeight: '40px',
        fontWeight: 300,
        fontFamily: '"Helvetica Neue", Helvetica, sans-serif',
        fontSize: '18px',
        '::placeholder': {
          color: '#CFD7E0'
        }
      }
    }
  };

  elementsOptions: ElementsOptions = {
    locale: 'en'
  };

  constructor(
    private _stripe: StripeService
  ) {}

  cardUpdated(result) {
    this.element = result.element;
    this.complete = result.card.complete;
    if(this.complete) {
      this._stripe.
    }
    this.error = undefined;
  }

  keyUpdated() {
    this._stripe.changeKey(this.stripeKey);
  }

  async payNow() {
    const response = await fetch('/paymentIntent', {
      method: 'POST',
      body: JSON.stringify({amount: 10000, currency: 'usd'}),
      headers: {'content-type': 'application/json'},
    });

    const body = await response.json();
    this._stripe.handleCardPayment(body.client_secret, this.element, { payment_method_data: {
      billing_details: {name: 'Bob Smith'}
    }}).subscribe(result => {
      if(result.error) {
        console.error('got stripe error', result.error);
      } else {
        console.log('payment succeeded');
      }
    });
  }
}
```

Please note you can also manually handle this flow using the paymentIntent methods however that is not the default path and you will have to implement it yourself (see more [here](https://stripe.com/docs/payments/3d-secure))

## Testing
The following command runs unit & integration tests that are in the `tests` folder, and unit tests that are in `src` folder:
```Shell
yarn test
```

## Building
The following command:
```Shell
yarn build
```
- starts _TSLint_ with _Codelyzer_
- starts _AoT compilation_ using _ngc_ compiler
- creates `dist` folder with all the files of distribution

To test the npm package locally, use the following command:
```Shell
yarn publish:dev
```
You can then run the following to install it in an app to test it:
```Shell
yalc link @nomadreservations/ngx-stripe
```

## Publishing

```Shell
yarn release:patch
or
yarn release:minor
or
yarn release:major
```

## Documentation
To generate the documentation, this starter uses [compodoc](https://github.com/compodoc/compodoc):
```Shell
yarn compodoc
yarn compodoc:serve
```

## License

MIT


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [1.2.3](https://github.com/nomadreservations/ngx-stripe/compare/1.2.2...1.2.3)

> 11 May 2020

- build(deps): bump https-proxy-agent from 2.2.2 to 2.2.4 [`#41`](https://github.com/nomadreservations/ngx-stripe/pull/41)
- build(deps): bump acorn from 5.7.3 to 5.7.4 [`#37`](https://github.com/nomadreservations/ngx-stripe/pull/37)
- feat(Payment Intends): Add support for confirmCardPayment method [`#34`](https://github.com/nomadreservations/ngx-stripe/pull/34)

#### [1.2.2](https://github.com/nomadreservations/ngx-stripe/compare/1.2.0...1.2.2)

> 30 January 2020

- Implement create payment method [`#23`](https://github.com/nomadreservations/ngx-stripe/pull/23)
- implement create payment method [`1a22131`](https://github.com/nomadreservations/ngx-stripe/commit/1a2213117adbdb0155cc4f084bac29af22b1e91d)
- filled in missing parts of type definitions. [`109fe4e`](https://github.com/nomadreservations/ngx-stripe/commit/109fe4e02880dc2fc0affc251dd298d801ff04a3)
- Release 1.2.2 [`c2caa4f`](https://github.com/nomadreservations/ngx-stripe/commit/c2caa4f8fe5dc4ab9d585e34ec0592b763af7ea6)

#### [1.2.0](https://github.com/nomadreservations/ngx-stripe/compare/1.2.0-beta.6...1.2.0)

> 4 December 2019

- #19: Fixes race condition on card view element [`0251b13`](https://github.com/nomadreservations/ngx-stripe/commit/0251b13081b9b7e9797cf7c1dbfc01f0c5f383d8)
- doc(#6): Document payment request button [`cf2eb26`](https://github.com/nomadreservations/ngx-stripe/commit/cf2eb2677c9d5516fc75c189efd0a51abd46c5bd)
- doc(#4): Payment intents and 3d Secure [`cb855f4`](https://github.com/nomadreservations/ngx-stripe/commit/cb855f413b81af2547900e60c77ac5983eb8df5e)

#### [1.2.0-beta.6](https://github.com/nomadreservations/ngx-stripe/compare/1.2.0-beta.5...1.2.0-beta.6)

> 8 October 2019

- chore(): Update for security issues [`5e58790`](https://github.com/nomadreservations/ngx-stripe/commit/5e587901cbc0278abb72d3137d9af42c58fcff83)
- Release 1.2.0-beta.6 [`25be609`](https://github.com/nomadreservations/ngx-stripe/commit/25be609d3d705b1fdb1ce267c5093c78b6322fd2)

#### [1.2.0-beta.5](https://github.com/nomadreservations/ngx-stripe/compare/1.2.0-beta.4...1.2.0-beta.5)

> 8 October 2019

- feet(#6): Implement request payment element, needs documentation [`1daf2d0`](https://github.com/nomadreservations/ngx-stripe/commit/1daf2d0961a751041a1aede3109422ff737a9e5c)
- Release 1.2.0-beta.5 [`ecc9fa4`](https://github.com/nomadreservations/ngx-stripe/commit/ecc9fa4a93a43219594095f5da00949ae97e2609)

#### [1.2.0-beta.4](https://github.com/nomadreservations/ngx-stripe/compare/1.2.0-beta.3...1.2.0-beta.4)

> 8 October 2019

- feet(#6): Implement request payment element, needs documentation [`fc44da7`](https://github.com/nomadreservations/ngx-stripe/commit/fc44da70e3381301209b3ee355e8de7609d0431d)
- feat(#16): Document factory [`742eaad`](https://github.com/nomadreservations/ngx-stripe/commit/742eaad040e42ba9af83269134b6721b89a1650a)
- Release 1.2.0-beta.4 [`0d8e723`](https://github.com/nomadreservations/ngx-stripe/commit/0d8e72305bef93b8d73c708f0a294941b5d0e4fd)

#### [1.2.0-beta.3](https://github.com/nomadreservations/ngx-stripe/compare/1.2.0-beta.2...1.2.0-beta.3)

> 7 October 2019

- feat(#4): Expose payment intent and 3d secure actions [`3a85185`](https://github.com/nomadreservations/ngx-stripe/commit/3a8518556fb4fb8397cac6336575e1482b8fbe50)
- Release 1.2.0-beta.3 [`400f146`](https://github.com/nomadreservations/ngx-stripe/commit/400f146813b6bdbaaf4b98225f912b8653088792)

#### [1.2.0-beta.2](https://github.com/nomadreservations/ngx-stripe/compare/1.2.0-beta.1...1.2.0-beta.2)

> 7 October 2019

- feat(#15): Adds ability to split elements [`a5d3af3`](https://github.com/nomadreservations/ngx-stripe/commit/a5d3af30dcab13523addcf6ec13e6c8279c93aef)
- Release 1.2.0-beta.2 [`24ce18e`](https://github.com/nomadreservations/ngx-stripe/commit/24ce18ef55a329d92e6abc45808caa70b43ff9a3)

#### [1.2.0-beta.1](https://github.com/nomadreservations/ngx-stripe/compare/1.2.0-beta.0...1.2.0-beta.1)

> 7 October 2019

- upgrade readme [`a34a014`](https://github.com/nomadreservations/ngx-stripe/commit/a34a0140449d93b09cbe4d38739cf8d5612cf662)
- upgrade angular [`3a4d632`](https://github.com/nomadreservations/ngx-stripe/commit/3a4d632c9c98bc7582dd3d8b29abf47f0987347d)
- Release 1.2.0-beta.1 [`300baf8`](https://github.com/nomadreservations/ngx-stripe/commit/300baf83321c259b78c96260a811b6e0f4c07b90)

#### [1.2.0-beta.0](https://github.com/nomadreservations/ngx-stripe/compare/1.1.0-beta.0...1.2.0-beta.0)

> 4 September 2019

- Release 1.2.0-beta.0 [`207f078`](https://github.com/nomadreservations/ngx-stripe/commit/207f078f005617ce5f6b48b0b888d591e2bbdb82)
- merged [`d573de5`](https://github.com/nomadreservations/ngx-stripe/commit/d573de5e36b1630e12022f1988f478fddf90896b)

#### [1.1.0-beta.0](https://github.com/nomadreservations/ngx-stripe/compare/1.0.1...1.1.0-beta.0)

> 4 September 2019

- Fixes currently fixable parts of #11, needs angular 8 to fix others [`#12`](https://github.com/nomadreservations/ngx-stripe/pull/12)
- Implement handleCardSetup() [`#9`](https://github.com/nomadreservations/ngx-stripe/pull/9)
- Fixes import example [`#2`](https://github.com/nomadreservations/ngx-stripe/pull/2)
- update to angular 7, make an angular-cli project [`5ec5114`](https://github.com/nomadreservations/ngx-stripe/commit/5ec5114ac6c260d7e6c64ac14b1bd3d49d48d3f6)
- update to angular 7, make an angular-cli project [`bf3528f`](https://github.com/nomadreservations/ngx-stripe/commit/bf3528f1f50cf337a64f45c6c9d788d10490c410)
- Update to angular 6 and rxjs 6 [`3c707fb`](https://github.com/nomadreservations/ngx-stripe/commit/3c707fbe2de6606abb99e1b6e41c089c9ad78c91)

#### [1.0.1](https://github.com/nomadreservations/ngx-stripe/compare/1.0.0...1.0.1)

> 21 January 2018


src/app/app.component.ts
import { AfterContentInit, Component } from '@angular/core';
import { FormControl, Validators } from '@angular/forms';
import { MatDialog } from '@angular/material/dialog';
import {
  Element as StripeElement,
  ElementOptions,
  ElementsOptions,
  StripeService
} from '@nomadreservations/ngx-stripe';
import { PaymentRequestButtonStyle, RequestElementOptions } from 'projects/ngx-stripe/src/lib/interfaces/element';
import { environment } from 'src/environments/environment';
import { DialogComponent } from './dialog/dialog.component';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css']
})
export class AppComponent implements AfterContentInit {
  error: any;
  stripeKey = new FormControl('', [Validators.required, validateStripeTestKey]);
  keyError = 'Must be a valid publishable test key (e.g. beings with pk_test_)';
  complete = false;
  complete2 = false;
  element: StripeElement;
  cardOptions: ElementOptions = {
    style: {
      base: {
        iconColor: '#276fd3',
        color: '#31325F',
        lineHeight: '40px',
        fontWeight: 300,
        fontFamily: '"Helvetica Neue", Helvetica, sans-serif',
        fontSize: '18px',
        '::placeholder': {
          color: '#CFD7E0'
        }
      }
    }
  };

  requestOptions: RequestElementOptions = {
    country: 'US',
    currency: 'usd',
    requestPayerName: true,
    requestPayerEmail: true,
    requestPayerPhone: true,
    requestShipping: true,
    total: {
      amount: 10000,
      label: 'Donate to the things'
    }
  };

  styles: PaymentRequestButtonStyle = {
    type: 'donate',
    theme: 'light',
    height: '64px'
  };

  elementsOptions: ElementsOptions = {
    locale: 'en'
  };

  public pay = 100;

  token: any = {};

  constructor(private _stripe: StripeService, private _dialog: MatDialog) {
    this.stripeKey.setValue('');
  }

  ngAfterContentInit() {
    this.keyUpdated();
  }

  cardUpdated(result) {
    console.log(result);
    this.element = result.element;
    this.complete = result.complete;
    this.error = undefined;
  }

  cardUpdated2(result) {
    console.log('update second card', result);
    this.element = result.element;
    this.complete2 = result.complete;
    this.error = undefined;
  }

  keyUpdated() {
    if (this.stripeKey.valid && this.stripeKey.value.length > 0) {
      this._stripe.changeKey(this.stripeKey.value);
    } else if (environment.stripeKey) {
      this._stripe.changeKey(environment.stripeKey);
    }
  }

  updatePay(amount: number) {
    this.pay = amount;
    this.requestOptions = {
      ...this.requestOptions,
      total: {
        ...this.requestOptions.total,
        amount: amount * 100
      }
    };
  }

  openPaymentDialog(): void {
    const dialogRef = this._dialog.open(DialogComponent, {
      width: '370px',
      data: { key: this.stripeKey }
    });
    dialogRef.afterClosed().subscribe(token => {
      this.token = token;
    });
  }

  requestUpdated(result) {
    console.log('updated token ', result);
    this.token = result;
    // if the server completes payment call this
    result.complete('success');
    // if the server fails to complete payment call this instead
    // result.complete('fail');
  }

  updateShippingAddress(result) {
    console.log('shipping address change', result);
    result.updateWith({
      status: 'success',
      shippingOptions: [
        {
          id: 'free-shipping',
          label: 'Free shipping',
          detail: 'Arrives in 5 to 7 days',
          amount: 0
        },
        {
          id: 'express',
          label: 'Express shipping',
          detail: 'Arrives in 1 to 2 days',
          amount: 1000
        }
      ]
    });
  }
  updateShippingOption(result) {
    console.log('shipping option change', result);
    result.updateWith({
      status: 'success'
    });
  }

  getCardToken() {
    this._stripe
      .createToken(this.element, {
        name: 'tested_ca',
        address_line1: '123 A Place',
        address_line2: 'Suite 100',
        address_city: 'Irving',
        address_state: 'BC',
        address_zip: 'VOE 1H0',
        address_country: 'CA'
      })
      .subscribe(result => {
        this.token = result;
      });
  }
}

function validateStripeTestKey(control: FormControl) {
  if (control.value && control.value.match(/^pk_test_.+/)) {
    return;
  }
  return { valid: false };
}
